### DISCUSSION TITLE
clsx__WEBPACK_IMPORTED_MODULE_5__ is not a function

### DISCUSSION THREAD

**Discussion by mistrya** [react-querybuilder/react-querybuilder#735] [Q&A]
Hello 
I am trying out react-querybuilder for first time.   After installing the package, I tried to use simple basic typescript as mentioned in https://react-querybuilder.js.org/docs/intro - 
-------- MyQueryBuider.tsx ------
```
import { QueryBuilder } from 'react-querybuilder';
import 'react-querybuilder/dist/query-builder.css';

export default () => <QueryBuilder/>; 
```
  -----
And referenced it from App.tsx 

When I launch the app I get below error.   Can you please help. 

react-querybuilder.mjs:2152 Uncaught TypeError: clsx__WEBPACK_IMPORTED_MODULE_5__ is not a function
    at react-querybuilder.mjs:2152:63
    at Array.map (<anonymous>)

----------
UPDATE - 07/11/2024
----------------
I installed several versions and issue persist on all versions 6.2.0 and above.    it worked fine in 6.1.4

**Comment by jakeboone02** [comment]
Can you share a minimal repo reproducing the issue?

**Reply by mistrya** [reply]
I will try to get a minimal repo .  
but wanted to share the error detail - 
TypeError
clsx__WEBPACK_IMPORTED_MODULE_5__ is not a function
Call Stack
 undefined
  vendors.js:295078:92
 joinClassnamesByName
  vendors.js:295078:81
 mergeClassnames
  vendors.js:295080:19
 undefined
  vendors.js:295481:11
 mountMemo
  vendors.js:238123:19
 Object.useMemo
  vendors.js:238568:16
 useMemo
  vendors.js:261328:21
 useMergedContext
  vendors.js:295480:60
 useQueryBuilderSetup
  vendors.js:296411:22
  
-----
From console - 
react-querybuilder.mjs:2152 Uncaught TypeError: clsx__WEBPACK_IMPORTED_MODULE_5__ is not a function
    at react-querybuilder.mjs:2152:63
    at Array.map (<anonymous>)
    at joinClassnamesByName (react-querybuilder.mjs:2152:46)
    at mergeClassnames (react-querybuilder.mjs:2154:1)
    at react-querybuilder.mjs:2555:1
    at mountMemo (react-dom.development.js:16406:1)
    at Object.useMemo (react-dom.development.js:16851:1)
    at useMemo (react.development.js:1650:1)
    at useMergedContext (react-querybuilder.mjs:2554:21)
    at useQueryBuilderSetup (react-querybuilder.mjs:3485:1)

**Reply by jakeboone02** [reply]
maybe just share your `package.json`

**Comment by mistrya** [comment]
{
  "name": "myproject-ui",
  "version": "0.0.0",
  "description": "",
  "keywords": [],
  "homepage": "",
  "bugs": {
    "url": "https://repoUrl",
    "email": "test@test.com"
  },
  "repository": {
    "type": "git",
    "url": "https://repoUrl"
  },
  "license": "UNLICENSED",
  "author": "",
  "files": [
    "dist"
  ],
  "scripts": {
    "analyze": "xxxxx-scripts webpack --mode=production --env=analyze",
    "build": "xxxxx-scripts webpack --mode=production",
    "build:dev": "xxxxx-scripts webpack --mode=development",
    "clean": "rm -rf coverage dist node_modules",
    "clean:dist": "rm -rf dist",
    "codegen": "codegen",
    "container:build": "docker build -t myproject-ui:latest .",
    "container:run": "docker run --rm -it -p 8080:8080 myproject-ui:latest",
    "coverage": "xxxxx-scripts jest --coverage",
    "dev": "xxxxx-scripts webpack serve --port=8080",
    "doctor": "xxxxx-scripts doctor",
    "eslint": "xxxxx-scripts eslint",
    "eslint:fix": "xxxxx-scripts eslint --fix",
    "lint": "concurrently --raw --group \"xxxxx-scripts eslint\" \"xxxxx-scripts stylelint\"",
    "lint:fix": "concurrently --raw --group \"xxxxx-scripts eslint --fix\" \"xxxxx-scripts stylelint --fix\"",
    "prepare": "xxxxx-scripts governance",
    "prettier": "xxxxx-scripts prettier . --check",
    "prettier:fix": "xxxxx-scripts prettier . --write",
    "serve": "xxxxx-scripts sirv dist --single --cors --dev",
    "serve:prod": "xxxxx-scripts sirv dist --single --host",
    "stylelint": "xxxxx-scripts stylelint",
    "stylelint:fix": "xxxxx-scripts stylelint --fix",
    "test": "xxxxx-scripts jest",
    "test:watch": "xxxxx-scripts jest --watch",
    "types": "tsc --noEmit"
  },
  "dependencies": {
    "@ag-grid-community/client-side-row-model": "^31.3.2",
    "@ag-grid-community/core": "^31.3.2",
    "@ag-grid-community/react": "^31.3.2",
    "@ag-grid-community/styles": "^31.3.2",
    "@mds/resources-web": "^15.4.0",
    "@mds/web": "^15.4.0",
    "@mds/web-illustrations": "^15.3.0",
    "@mds/web-ui-button": "~2.0.0",
    "@mds/web-ui-icons": "~2.0.0",
    "@mds/web-ui-link": "~2.0.0",
    "@mds/web-ui-spinner": "~2.0.0",
    "@mds/web-ui-text": "~2.0.1",
    "@mds/web-ui-theme": "~2.0.0",
    "@popperjs/core": "^2.11.8",
    "@t3-oss/env-core": "~0.8.0",
    "@yyyyy/bootstrap-theme": "^13.0.0",
    "@yyyyy/react-component-library": "^13.0.0",
    "@yyyyy/tokens": "^13.0.0",
    "bootstrap": "^4.6.2",
    "clsx": "~2.1.1",
    "lodash": "^4.17.21",
    "react": "^18.3.1",
    "react-app-polyfill": "^2.0.0",
    "react-beautiful-dnd": "^13.1.1",
    "react-datepicker": "^4.10.0",
    "react-dom": "^18.3.1",
    "react-inlinesvg": "^3.0.2",
    "react-popper": "^2.3.0",
    "**react-querybuilder": "^7.5.2**",
    "react-router-dom": "^6.23.1",
    "reactstrap": "^8.10.1",
    "single-spa-react": "~6.0.1",
    "zod": "~3.23.8"
  },
  "devDependencies": {
    "@axe-core/react": "~4.9.0",
    "@xxxxx/browserslist-config": "~1.2.1",
    "@xxxxx/lefthook-config": "~0.1.1",
    "@xxxxx/scripts": "^1.27.0",
    "@testing-library/dom": "~9.3.4",
    "@testing-library/react": "~14.3.1",
    "@testing-library/user-event": "~14.5.2",
    "@total-typescript/ts-reset": "~0.5.1",
    "@types/estree": "~1.0.5",
    "@types/jest": "~29.5.12",
    "@types/jest-axe": "~3.5.9",
    "@types/node": "~20.10.8",
    "@types/react": "~18.3.2",
    "@types/react-dom": "~18.3.0",
    "@types/testing-library__jest-dom": "~5.14.9",
    "concurrently": "~8.2.2",
    "cross-env": "~7.0.3",
    "cross-fetch": "~4.0.0",
    "dotenv-webpack": "~8.1.0",
    "jest-axe": "~8.0.0",
    "lefthook": "~1.6.11",
    "msw": "~2.3.0",
    "react-test-renderer": "~18.3.1",
    "rimraf": "~5.0.7",
    "single-spa": "~6.0.1",
    "type-fest": "~4.18.2",
    "typescript": "~5.4.5",
    "webpack-merge": "~5.10.0"
  },
  "browserslist": [
    "extends @xxxxx/browserslist-config"
  ],
  "engines": {
    "node": ">=20"
  },
  "readme": "./README.md"
}

**Reply by mistrya** [reply]
I replaced all two internal libs with xxxxx and yyyyy

**Reply by jakeboone02** [reply]
Thanks. What version of webpack is used in `@xxxxx/scripts`?

**Reply by mistrya** [reply]
So we have wrapper around webpack and that wrapper library has below webpack dependencies in package.json

  "devDependencies": {
    "@swc/cli": "~0.4.0",
    "@types/webpack-env": "~1.18.5",
    "eslint": "~8.57.0",
    "jest": "~29.7.0",
    "prettier": "~3.3.2",
    "webpack": "~5.92.1",
    "webpack-cli": "~5.1.4",
    "webpack-dev-server": "~4.15.2",
    "wireit": "~0.14.5",
    "@xxxxxrepo/tsconfig": "0.0.0"
  },
  "peerDependencies": {
    "webpack": "~5.91.0",
    "webpack-cli": "~5.1.4",
    "webpack-dev-server": "~4.15.2"
  },

**Reply by jakeboone02** [reply]
Hmmm...I thought maybe you were on an old webpack version but that's very recent. I think I'll have to see a repo that reproduces the problem to figure out what's happening.

**Reply by mistrya** [reply]
I will try to create one and share

**Reply by jaganryali** [reply]
Hey @jakeboone02 ,

I'm from the same team that @mistrya is from. 
Just want to add that we noticed a difference that we are facing this issue in the app where there is singleSpa and Module Federation is included. We don't have the issue in the repo where we don't have singleSpa and Module Federation.

**Reply by jakeboone02** [reply]
I'm not familiar with those, are they webpack configs?

**Reply by jaganryali** [reply]
yeah to build micro frontends

**Reply by jakeboone02** [reply]
OK. I would definitely need to see a working (well, you know what I mean) reproduction.
